h/api/search/transform.py

```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
+
+"""Functions for transforming data for or from the search index."""
+
+import copy
+
+from h.api import uri
+
+
+def prepare(annotation):
+    """
+    Prepare the given annotation for indexing.
+
+    Scan the passed annotation for any target URIs or document metadata URIs
+    and add normalised versions of these to the document.
+    """
+
+    # FIXME: When this becomes simply part of a search indexing operation, this
+    # should probably not mutate its argument.
+    _normalise_annotation_target_uris(annotation)
+
+
+def render(annotation):
+    """
+    Render an annotation retrieved from search for public display.
+
+    Receives data direct from the search index and reformats it for rendering
+    or display in the public API. At the moment all this does is remove
+    normalised URI fields.
+    """
+    data = copy.deepcopy(annotation)
+
+    _filter_target_normalised_uris(data)
+
+    return data
+
+
+def _normalise_annotation_target_uris(annotation):
+    if 'target' not in annotation:
+        return
+    if not isinstance(annotation['target'], list):
+        return
+    for target in annotation['target']:
+        if not isinstance(target, dict):
+            continue
+        if 'source' not in target:
+            continue
+        if not isinstance(target['source'], basestring):
+            continue
+        target['source_normalised'] = uri.normalise(target['source'])
```

Comment:
Is there a reason we can't just exclude this from the source in the mapping? That would let it be searchable and indexed but not stored in the source. It means that for a re-index we'd have to run annotations through our Python code again to regenerate these things, but shouldn't we do that anyway if we're starting to handle things like this in Python?